ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosChangeFeed.java

```diff
@@ -0,0 +1,188 @@
+   * The number of entries is capped by maxEntriesSize.
+   * This method creates a cache for change feed entries. If the {@code azureFindToken} is not valid,
+   * or if all the items in the cache are consumed, then it queries cosmos for new entries.
+   * @param azureFindToken {@link AzureFindToken} after which the next entries have to be returned.
+   * @param results {@link List} of {@link CloudBlobMetadata} objects which will be populated by new entries.
+   * @param maxEntriesSize maximum size of all the blobs returned in {@code results}
+   * @param partitionId Partition for which change feed entries have to be returned.
+   * @return updated {@link AzureFindToken} after processing the next set of entries.
+   */
+  public AzureFindToken getNextEntriesAndToken(AzureFindToken azureFindToken, List<CloudBlobMetadata> results,
+      long maxEntriesSize, String partitionId) throws DocumentClientException {
+    int index = azureFindToken.getIndex();
+    if (!changeFeedCache.containsKey(partitionId) || !isCacheValid(partitionId, azureFindToken)) {
+      populateChangeFeedCache(partitionId, azureFindToken.getStartContinuationToken());
+      index = 0;
+    }
+
+    long resultSize = 0;
+    while (true) {
+      if (index < changeFeedCache.get(partitionId).getFetchedEntries().size()) {
+        if (resultSize + changeFeedCache.get(partitionId).getFetchedEntries().get(index).getSize() < maxEntriesSize) {
+          results.add(changeFeedCache.get(partitionId).getFetchedEntries().get(index));
+          resultSize = resultSize + changeFeedCache.get(partitionId).getFetchedEntries().get(index).getSize();
+          index++;
+        } else {
+          if (resultSize == 0) {
+            results.add(changeFeedCache.get(partitionId).getFetchedEntries().get(index));
+            index++;
+          }
+          break;
+        }
+      } else {
+        populateChangeFeedCache(partitionId, azureFindToken.getEndContinuationToken());
+        if (cacheEmpty(partitionId)) {
+          // this means that there are no new changes
+          break;
+        }
+        index = 0;
+      }
+    }
+
+    return new AzureFindToken(changeFeedCache.get(partitionId).getStartContinuationToken(),
+        changeFeedCache.get(partitionId).getEndContinuationToken(), index,
+        changeFeedCache.get(partitionId).getFetchedEntries().size(),
+        changeFeedCache.get(partitionId).getAzureRequestId());
+  }
+
+  /**
+   * Check is the cache is valid for the {@code azureFindToken} provided.
+   * @param partitionId partition of the {@code azureFindToken}.
+   * @param azureFindToken {@link AzureFindToken} object.
+   * @return true is cache is valid. false otherwise.
+   */
+  private boolean isCacheValid(String partitionId, AzureFindToken azureFindToken) {
+    ChangeFeedCacheEntry changeFeedCacheEntry = changeFeedCache.get(partitionId);
+    return Utils.checkNullableStringEquals(azureFindToken.getAzureTokenRequestId(),
```

Comment:
Don't you have an equals method that does the same check?

Reply:
this is not exactly same as equals, as it can have a different index. Semantically this is a different test than equals. It tests that the cache is still valid for the change feed token in request.